de.metas.contracts/src/main/java/de/metas/contracts/impl/ContractChangeBL.java

```diff
@@ -86,31 +86,41 @@ public void cancelContract(@NonNull final I_C_Flatrate_Term currentTerm,
 			cancelContractIfNotCanceledAlready(initialContract, contractChangeParameters);
 		}
 	}
-	
+
+	private boolean isVoidSingleContract(final @NonNull ContractChangeParameters contractChangeParameters)
+	{
+		return ChangeTerm_ACTION_VoidSingleContract.equals(extractAction(contractChangeParameters));
+	}
+
+	private String extractAction(final @NonNull ContractChangeParameters contractChangeParameters)
+	{
+		return contractChangeParameters.getAction();
+	}
+
 	private void unlinkContractIfNeeded(@NonNull final I_C_Flatrate_Term currentTerm,
 			final @NonNull ContractChangeParameters contractChangeParameters)
 	{
 		final I_C_Flatrate_Term ancestor = flatrateDAO.retrieveAncestorFlatrateTerm(currentTerm);
-		 if (ancestor != null && contractChangeParameters.isOnlyTerminateCurrentTerm())
-		 {
-			 ancestor.setC_FlatrateTerm_Next(null);
-			 ancestor.setAD_PInstance_EndOfTerm(null);
-			 setAncestorMasterEndDateWhenUnlinkContract(ancestor);
-			 
-			 InterfaceWrapperHelper.save(ancestor);
-		 }
+		if (ancestor != null && isVoidSingleContract(contractChangeParameters))
```

Comment:
optimization hint: call isVoidSingleContract(contractChangeParameters) before loading "ancestor".
If that method returns false you can return from this method directly (shortcut), without loading that ancestor.

Reply:
thx!